### ISSUE TITLE
create server credentials

### ISSUE THREAD

**Issue by schoripaz** [jenkinsci/docker-swarm-plugin#35]
hi jenkinsci
i am working with jenkins on standalone docker host, in addition i am having docker swarm with 4 hosts.
my jenkins configured with docker-swarm-plugin, all works fine when using none secure -H=0.0.0.0:2375.
i manage to configure my docker swarm host using tls on port H=0.0.0.0:2376 as describe in this artificial:
https://docs.docker.com/engine/security/https/
but could not manage to connect it from jenkins docker-swarm-plugin can you add some more details on how to create "Server credentials" as this image shows?
https://github.com/jenkinsci/docker-swarm-plugin/blob/master/docs/images/configuration.png 

Thanks Paz

**Comment by Roemer** [comment]
The server credentials are very normal Jenkins credentials. For this, see https://jenkins.io/doc/book/using/using-credentials/

**Comment by pomrw** [comment]
Can you elobrate on the "very normal Jenkins credentials" please.
My Jenkins setup has several credentials of various sorts (username/password, certificates, secret text, etc.) but the "Server credentials" selection box in the Docker Swarm Cloud Configuration pane only contains "none".
Trying to add credentials with the "Add" button next to it does not change anything.

The "Test Connection" button results in stack traces.

I can't get this to work.
A little more documentation or at least help for diagnosing the problem would be helpful.

**Comment by gutschet** [comment]
I have the same problem.
What do I have to enter for Key, Cert and CA? The file-path? The file content?

**Comment by maacarbo** [comment]
I fully agree that some more documentation about this topic would be _more_ than welcome!

**Comment by David-Villeneuve** [comment]
... and where do you get the key, cert and ca.   I've been scouring the net looking for this. Clearly nobody is using this plug-in

**Comment by gutschet** [comment]
It is the CA, Key and Crt that you use to TLS secure the docker daemon of your swarm manager. In my case, my dockerd start command looks like this:
```
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --tlsverify --tlscacert /my/path/docker-ca.pem --tlscert /my/path/docker-master.domain.com.crt --tlskey /my/path/docker-master.domain.com.key-H tcp://0.0.0.0:2376
```
And these are the ones you need to use in server credentials.+

By the way: we are using the plugin in production for a 10 node swarm cluster. In the cluster, we are running our complete devops tooling (jenkins, nexus, keycloak, ELK Stack, ...) and some of our applications test instances where our built applications are auto-deployed. We are running a second 3 node swarm cluster with the plugin, providing public available test-applications to customers.

The plugin is not really widely used but it works really well.

Kind regards

**Comment by David-Villeneuve** [comment]
Gutschet, I'm afraid you'll have to walk me through some details as to how you produce each certificate. Certainly the combination I've been trying refuses to work.  

The other thing is that the x509 settings wants a Client Key, not a server key, but no idea what this is.  It also want a client certificate, not the server certificate.  Then it wants a server CA Certificate, which I assume is the CA cert used to creating the server certificate and a client certificate.  

If you could post a clear concise step-by-step recipe from start to finish, then I think this plug-in would have more take up/support.  I'm sure a lot of people have given up on it like me from not being able to get it to work.

**Comment by gutschet** [comment]
@David-Villeneuve , look at that documentation please and then tell me, if you have any more questions

**Comment by dmckee-ebsco** [comment]
> Can you elaborate on the "very normal Jenkins credentials" please. My Jenkins setup has several credentials of various sorts (username/password, certificates, secret text, etc.) but the "Server credentials" selection box in the Docker Swarm Cloud Configuration pane only contains "none". Trying to add credentials with the "Add" button next to it does not change anything (i.e. a newly created credential does not show up in the list).
> 
> The "Test Connection" button results in stack traces.
> 
> I can't get this to work. A little more documentation or at least help for diagnosing the problem would be helpful.

(Disclaimer: I am using the Docker plugin, not swarm)
If you are having the exact issue that @pomrw is describing here, leave this server credential drop-menu as `none`, scroll down and expand the `docker agent templates` and set `connection method` to "Connect with SSH" and "Use configured SSH credentials. Your credentials should be listed here.
